### PR TITLE
[IGA] Support Nitsche Condition

### PR DESCRIPTION
**Pull request by rickyaristio**
**Description**
This PR adds the new Nitsche-type support condition. The implementation is quite similar to #8827, but much simpler. 

**Changelog**
Additional features added to obtain the Nitsche stabilization factor for both coupling and support Nitsche condition (see nitsche_stabilization_process.py and eigensolver_nitsche_stabilization_strategy.h).

### CODE REVIEW

File: applications/IgaApplication/custom_conditions/support_nitsche_condition.h [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+        /* Transformation the strain tensor from the curvilinear system
+        *  to the local cartesian in voigt notation including a 2 in the
+        *  shear part. */
+        std::vector<Matrix> m_T_vector;
+
+        /* Transformation the stress tensor from the local cartesian 
+        *  to the curvilinear system in voigt notation. */
+        std::vector<Matrix> m_T_hat_vector;
+
+        std::vector<array_1d< array_1d<double, 3>,2>> m_reference_contravariant_base;
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_conditions/support_nitsche_condition.h:330]: comment?

File: applications/IgaApplication/custom_conditions/support_nitsche_condition.h [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+            const KinematicVariables& rActualKinematic,
+            ConstitutiveVariables& rThisConstitutiveVariablesMembrane);
+
+        void CalculateSecondVariationTractionProduct(
+            IndexType IntegrationPointIndex,
+            Matrix& rPi,
+            const KinematicVariables& rActualKinematic,
+            ConstitutiveVariables& rThisConstitutiveVariablesMembrane);
+
+        void CalculateSecondVariationTraction(
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_conditions/support_nitsche_condition.h]: would be good to have comments for those functions

File: applications/IgaApplication/custom_conditions/support_nitsche_condition.h [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+            Matrix& rFirstVariationStressCovariant, 
+            array_1d<double, 3>& rDisplacement,
+            array_1d<double, 3>& rSecondVariationTractionProduct);
+        
+        /**
+        * This functions updates the constitutive variables
+        * @param rActualMetric: The actual metric
+        * @param rThisConstitutiveVariables: The constitutive variables to be calculated
+        * @param rValues: The CL parameters
+        * @param ThisStressMeasure: The stress measure considered
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_conditions/support_nitsche_condition.h]: parameters don't match

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 4]
```diff
@@ -389,95 +389,161 @@ class EigensolverNitscheStabilizationStrategy
...
+            // 1. find the DOFs on the current interface boundary
+            Model reduced_model_master;               
+            ModelPart& reduced_model_part_master = reduced_model_master.CreateModelPart("new_model"); 
 
-        for (auto& r_cond : rModelPart.Conditions()) {
-            
-            auto& r_geom_master = r_cond.GetGeometry().GetGeometryPart(0);
-            auto& r_N_master = r_geom_master.ShapeFunctionsValues();
+            for (auto& r_cond : rModelPart.Conditions()) {
+
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp]:
```suggestion
```

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 5]
```diff
@@ -389,95 +389,161 @@ class EigensolverNitscheStabilizationStrategy
...
+            IndexType i_master = 0;
+            for (auto& r_node : reduced_model_part_master.Nodes()) {
 
-            const IndexType index = i_master * 3;
-            
-            rResult[index]     = r_node.GetDof(DISPLACEMENT_X).EquationId();
-            rResult[index + 1] = r_node.GetDof(DISPLACEMENT_Y).EquationId();
-            rResult[index + 2] = r_node.GetDof(DISPLACEMENT_Z).EquationId();
+                const IndexType index = i_master * 3;
+
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp:439]:
```suggestion

```

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 6]
```diff
@@ -389,95 +389,161 @@ class EigensolverNitscheStabilizationStrategy
...
+            reduced_stiffness_matrix = ZeroMatrix((number_of_nodes_master+number_of_nodes_slave)*3,(number_of_nodes_master+number_of_nodes_slave)*3);
 
-        for (IndexType i = 0; i < (number_of_nodes_master+number_of_nodes_slave)*3; i++) 
+            for (IndexType i = 0; i < (number_of_nodes_master+number_of_nodes_slave)*3; i++) 
+            {
+                for (IndexType j = 0; j <= i; j++) 
+                {
+                    reduced_stiffness_matrix(i,j) = rStiffnessMatrix(rResult(i),rResult(j));
+                    if (i != j)
+                        reduced_stiffness_matrix(j,i) = rStiffnessMatrix(rResult(i),rResult(j));
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp]:
```suggestion
                    if (i != j) {
                        reduced_stiffness_matrix(j,i) = rStiffnessMatrix(rResult(i),rResult(j));
                    }
```

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 7]
```diff
@@ -389,95 +389,161 @@ class EigensolverNitscheStabilizationStrategy
...
+            // 3. assigned the value of the reduced stifness and stabilization matrices based on result vector
+            reduced_stabilization_matrix = ZeroMatrix((number_of_nodes)*3,(number_of_nodes)*3);
+
+            for (IndexType i = 0; i < (number_of_nodes)*3; i++) 
+            {
+                for (IndexType j = 0; j <= i; j++) 
+                {
+                    reduced_stabilization_matrix(i,j) = rStabilizationMatrix(rResult(i),rResult(j));
+                    if (i != j)
+                        reduced_stabilization_matrix(j,i) = rStabilizationMatrix(rResult(i),rResult(j));
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp:531]:
```suggestion
                    if (i != j) {
                        reduced_stabilization_matrix(j,i) = rStabilizationMatrix(rResult(i),rResult(j));
                    }
```

try to always keep the brackets. It is also somewhere KRATOS-style guide.

**Comment by tteschemacher** [comment]
It looks really good. Just some really minor things.

**Comment by rickyaristio** [comment]
> It looks really good. Just some really minor things.

Thanks! Changes will be ready soon.

**Review by tteschemacher — APPROVED** [review]
looks fine to me!

Thanks a lot!

**Comment by rickyaristio** [comment]
> looks fine to me!
> 
> Thanks a lot!

You're welcome!

**Pull request merged**